### ISSUE TITLE
New .dockerignore pattern matcher is under-ignoring

### ISSUE THREAD

**Issue by LouisAmon** [pulumi/pulumi-docker#607]
### What happened?

I have the following folder structure:

```
projects/services/src
projects/api/src
projects/infrastructure/src
```

And then inside each project I have a `.venv` folder containing my Python dependencies:
```
projects/services/.venv
projects/api/.venv
projects/infrastructure/.venv
```

My .dockerignore is like this:
```
# Ignore everything
*

# Except these
!pyproject.toml
!poetry.lock

!projects/*/src
```

When I run `pulumi up`, I get the following error:
```
Diagnostics:
  docker:index:Image (services):
    error: unable to hash build context: open /workspace/py/projects/infrastructure/.venv/lib/python3.10/site-packages/IPython: no such file or directory
```

(My context and Dockerfile are both in /workspace/py)

### Expected Behavior

When I run `docker buildx build ...` then the image is built just fine, and if I explore inside with bash I can navigate that the correct files are there (and only those)

The `.venv` shouldn't even be considered by the hashing function, as it's ignored by the very first pattern `*`

### Steps to reproduce

I think we need a unit test for a more complex monorepo structure, because there's still some discrepancy between Docker's context hash and Pulumi's

### Output of `pulumi about`

CLI          
Version      3.64.0
Go Version   go1.20.3
Go Compiler  gc

Plugins
NAME       VERSION
aws        5.37.0
awsx       1.0.2
docker     4.2.0
python     unknown
random     4.13.0
snowflake  0.23.0

Host     
OS       debian
Version  11.6
Arch     aarch64

This project is written in python: executable='/usr/local/python/current/bin/python3' version='3.10.10

### Additional context

@guineveresaenger @AaronFriel this might be connected to the 4.2.0 release and your recent PR (https://github.com/pulumi/pulumi-docker/pull/595)

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

**Comment by guineveresaenger** [comment]
Hi @LouisAmon - thank you for following up.

Can you post a (simplified) file tree of your project files? I'm also curious whether the file that cannot be found does in fact exist (even though it should be ignored).

**Comment by guineveresaenger** [comment]
Hi @LouisAmon - 

so we did a deep dive on this, and here's the situation:

In a nutshell, [this behavior has been an issue for a long time](https://github.com/moby/moby/issues/45284). The reason it's working correctly when you run the CLI is that the docker CLI is now defaulting to using buildx, which relies on buildkit libraries.

The docker client has a buildkit mode, which we are using, but there are some really significant discrepancies that aren't fully implemented into the docker client yet, and this happens to be one of them.

We do plan to bring the behavior of our provider up to par with this functionality as soon as we can. In the meantime, we recommend avoiding wildcards in your `.dockerignore` paths.

**Comment by AaronFriel** [comment]
@LouisAmon It appears #596 should have fixed this, perhaps inadvertently by fixing path handling. In case it hasn't, #646 modifies our dockerignore handling to use the same directory crawl as upstream Docker's.

I began by modifying our implementation to use `fsutil.Walk`, the same library Docker Buildkit uses, set up a test case with wildcard exclusion rules (`!foo/*/bar`), and saw the tests pass. On backing out the change, the tests continued passing.

I checked out your repository - https://github.com/LouisAmon/pulumi-docker-4-bug and it appears that this should work on Docker latest.

**Comment by LouisAmon** [comment]
Awesome ! Looking forward to testing it 

Although, as per Guin’s last comment, I’ve refactored our monorepo to use a base image and then several docker files in each package : this means a very simplified context, and definitely no wildcard 😉

I wish I could go back to whitelisting select files though : it’s safer than blacklisting because you could always forget some config file which, in turn, invalidates your cache 


tl;dr

Looking forward to the release, thanks 🙏
